packages/service-core/src/sync/BucketChecksumState.ts

```diff
@@ -63,6 +63,15 @@ export class BucketChecksumState {
     }
   }
 
+  /**
+   * Build a new checkpoint line for an underlying storage checkpoint update if any buckets have changed.
+   *
+   * When this returns null, the call is idempotent (meaning that no internal state is updated until a line
+   * is generated).
```

Comment:
Is there a way to restructure this function so we can better "guarantee" this behavior? E.g. splitting it into one idempotent function, and separate one to update the state.

I went through the function again, and there is quite a lot happening before null is returned, including updating `this.bucketDataPositions`, which feels risky.